contracts/gol.cairo

```diff
@@ -0,0 +1,450 @@
+        cell_index : felt
+    ):
+    # This does not trigger an evolution. Multiple give_life
+    # operations may be called, building up a shape before
+    # a turn triggers evolution.
+    alloc_locals
+    let (caller) = ensure_user()
+    let (generation, current_game_state) = verify_and_get_for_cell_activation(
+        cell_index=cell_index,
+        user=caller
+    )
+    pay(user=caller, credit_requirement=GIVE_LIFE_CREDIT_REQUIREMENT)
+    activate_cell(
+        generation=generation,
+        caller=caller,
+        cell_index=cell_index,
+        current_state=current_game_state
+    )
+
+    return ()
+end
+
+##########
+
+# Returns a felt of encoded cells for the specified game's generation.
+@view
+func view_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt,
+        generation : felt
+    ) -> (
+        generation_owner : felt,
+        game_state : felt
+    ):
+
+    let (game_info) = stored_game.read(game_id, generation)
+    return (game_info[0], game_info[1])
+end
+
+# Returns the current generation of a given game.
+@view
+func get_current_generation{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt
+    ) -> (
+        generation : felt
+    ):
+    let (generation) = current_generation.read(game_id)
+    return (generation)
+end
+
+##########
+
+func pay{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt, 
+        credit_requirement : felt
+    ):
+    ERC20._burn(user, Uint256(credit_requirement, 0))
+    return ()
+end
+
+func reward_user{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt
+    ):
+    ERC20._mint(user, Uint256(1, 0))
+    return ()
+end
+
+func ensure_user{
+        syscall_ptr : felt*
+    }(
+    ) -> (
+        caller : felt
+    ):
+    let (caller) = get_caller_address()
+    with_attr error_message(
+        "User not authenticated"
+    ):
+        assert_not_zero(caller)
+    end
+    return (caller)
+end
+
+func verify_new_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game : felt
+    ):
+    alloc_locals
+    let (local game_info) = stored_game.read(
+        game_id=game, generation=0)
+    let game_id = game_info[1]
+
+    with_attr error_message(
+        "Game with genesis state {game} already exists"
+    ):
+        assert_not_equal(game_id, game)
+    end
+
+    let (high, _) = split_felt(game)
+    let (invalid_bits, _) = unsigned_div_rem(high, 2**97)
+    with_attr error_message(
+        "Game {game} is not a valid game"
+    ):
+        assert invalid_bits = 0
+    end
+    return ()
+end
+
+func verify_and_get_for_cell_activation{
```

Comment:
Usually `and` in function name is a sign it can be simplified. In this case it can be split into: `get_last_state` and `assert_valid_cell_index`.